### PR TITLE
451006 - AliaksandrVelichka

### PR DESCRIPTION
**Pull request by Engelstrom**

### CODE REVIEW

File: Tracer/TracerLib/Formatters/ConsoleResultFormatter.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                {
+                    PrintMethodResults(child, nesting);
+                    Console.WriteLine($"{tab}{tab}*****************************");
+                }
+            }
+        }
+
+        private static string GetTab(int nesting)
+        {
+            var tabs = "";
```
Comment (RamanBut-Husaim) [inline Tracer/TracerLib/Formatters/ConsoleResultFormatter.cs]: @Engelstrom use `string.Empty`.

File: Tracer/TracerLib/Formatters/ConsoleResultFormatter.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                    PrintMethodResults(child, nesting);
+                    Console.WriteLine($"{tab}{tab}*****************************");
+                }
+            }
+        }
+
+        private static string GetTab(int nesting)
+        {
+            var tabs = ""; 
+            for (int i = 0; i < nesting; i++)
```
Comment (RamanBut-Husaim) [inline Tracer/TracerLib/Formatters/ConsoleResultFormatter.cs]: @Engelstrom replace with `StringBuilder` or just use `new string(" ", nesting)`

File: Tracer/TracerLib/Formatters/XmlResultFormatter.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+            foreach (var Id in threads.Keys)
+            {
+                var thread = new XElement("thread", new XAttribute("id", Id));
+                PrintMethodResults(thread, threads[Id].HeadNode);
+                root.Add(thread);
+            }
+
+            document.Add(root);
+            document.Save("E:\\info.xml");
```
Comment (RamanBut-Husaim) [inline Tracer/TracerLib/Formatters/XmlResultFormatter.cs]: @Engelstrom move to 'private readonly' instance field

File: Tracer/TracerLib/Tracer.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+		private TraceResult result;
+
+		private Tracer()
+		{
+			result = new TraceResult();
+		}
+
+		public static Tracer GetInstance()
+		{
+			if (instance == null)
```
Comment (RamanBut-Husaim) [inline Tracer/TracerLib/Tracer.cs]: @Engelstrom Take a look at other possible [singleton implementations](http://csharpindepth.com/Articles/General/Singleton.aspx)

File: Tracer/TracerLib/Tracer.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+			}
+
+			return instance;
+		}
+
+		public void StartTrace()
+		{
+			lock (startMethodSync)
+			{
+				var stackTrace = new StackTrace();
```
Comment (RamanBut-Husaim) [inline Tracer/TracerLib/Tracer.cs]: @Engelstrom Could be moved out of lock.

**Comment by RamanBut-Husaim** [comment]
@Engelstrom All other issues were discussed personally.

**Comment by Engelstrom** [comment]
Issues have been fixed.

File: Tracer/TracerLib/Formatters/ConsoleResultFormatter.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+        public void Format(ImmutableDictionary<int, ThreadDescriptor> results)
+        {
+            throw new NotImplementedException();
+        }
+
+        private static string GetTab(int nesting)
+        {
+            var tabs = string.Empty; 
+            for (int i = 0; i < nesting; i++)
```
Comment (RamanBut-Husaim) [inline Tracer/TracerLib/Formatters/ConsoleResultFormatter.cs]: @Engelstrom Something strange happens in a loop.

File: Tracer/TracerLib/Formatters/XmlResultFormatter.cs [hunk 7]
```diff
@@ -0,0 +1,48 @@
+﻿using System.Xml.Linq;
+using TracerLib.Interfaces;
+using TracerLib.Utils;
+
+namespace TracerLib.Formatters
+{
+    public class XmlResultFormatter : ITraceResultFormatter
+    {
+        private readonly string fileName = "E:\\info.xml";
```
Comment (RamanBut-Husaim) [inline Tracer/TracerLib/Formatters/XmlResultFormatter.cs]: @Engelstrom Modify your formatter to accept a file path as a constructor parameter.

File: Tracer/TracerLib/Utils/TraceResult.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+
+namespace TracerLib.Utils
+{
+    public class TraceResult
+    {
+        public Dictionary<int, ThreadDescriptor> results { get; }
```
Comment (RamanBut-Husaim) [inline Tracer/TracerLib/Utils/TraceResult.cs]: @Engelstrom `PascalCase`

**Comment by RamanBut-Husaim** [comment]
@Engelstrom Accepted. Thank you.

**Pull request merged**